### DISCUSSION TITLE
Document metadata restrictions

### DISCUSSION THREAD

**Discussion by marcuslindblom** [ravendb/ravendb#14783] [Q&A]
Are there any known restrictions when storing data as metadata on a document or can I store any complex object and nested complex objects?

**Comment by garayx** [comment]
Hi

metadata is meant to be used by RavenDB for internal document properties such as collection name or last modified date.

You can store simple data there as string or numbers, and store object using `DynamicJsonValue`  or  `ExpandoObject`.

**Reply by marcuslindblom** [reply]
So when you say meant for, you mean, you should be careful to add own stuff as document metadata? Is that true @ayende

**Reply by ayende** [reply]
You _can_ store complex data there.
The question is what exactly are you trying to do?

The intend is to put stuff there that is generally cross cutting.

Note that we reserve all values with `@` prefix to RavenDB.

**Reply by marcuslindblom** [reply]
We store metadata about a page as document metadata. Like OG metadata, page title, page description and so on. When I say page I mean an object that is reached using a url.

**Reply by ayende** [reply]
Sure, that would work. but why is that in the metadata? Isn't that part of the doc's data directly?

**Reply by marcuslindblom** [reply]
Maybe it is... I can't really decide what I think is best. With the approach stuff that I would say is metadata is stored as metadata on the document. The feels correct to me 🤷‍♂️

**Reply by krivulcik** [reply]
I like to think about the metadata that it's a kind of "overhead" data necessary to make the document work. The system stores the additional information about the last update, replication, etc., there.

The *page* metadata is also a kind of "overhead" data, but not for the document. It's for the web page itself. So in your case, I would put it next to the page content, inside the document.

Also, consider a scenario where the page contains an image in the content, and you need to store some metadata for the image (e.g., dimensions, description, whatever). I don't think it would be best to store the *image* metadata in the *document* metadata space. Similarly, the page metadata should be kept separate from the document metadata. It's just data like any other from the database point of view.

**Reply by ayende** [reply]
Consider something like this:

```
{
   "Title": "Home Page",
    "Slugs": ["/", "/index", "/index.php"],
    "Body": [ "paragraph one", "para 2", "etc..." ],
    "Sections": [
            {"Type": "VisitorsCounter", "Location": "TopLeft" },
            {"Type": "Ad", "Location": "Middle" },
            {"Type": "Poll", "Location": "Bottom", "PollId": "polls/123-C" },
    ],
    "LastUpdatedBy": "Admin",
    "Tags": ["home", "index"],
     "@metadata": {
          "@collection": "Pages",
          "@last-modified": "2022-08-..."
      }
}
```


Does that represent a similar structure to what you have?

In this case, `Body` and `Section` are page "data", and things like `Title`, `LastUpdatedBy` are metadata. `Tags` I'm not sure how to classify.

But he key here is that all of those are application level entities.

**Reply by marcuslindblom** [reply]
Ok I see. One thing I struggle with is that my pages has not only slugs, it has an hierarchical url `/page-1/page-2`. The way I do it now is having one document with the url and the id to the document that should answer to this specific url. That works pretty great but has one huge backside. `session.Query<Post>()` for example gives me all posts for my blog but the Post does not have the url 🤯. 

I would love some to get some best practice into having a `Page` class or record with Slug and Url where the Url is hierarchical.
With your document in mind @ayende, how would you do it?

This question turned into something hole different 😀

**Reply by ayende** [reply]
@marcuslindblom I would love to do a session on this, would you mind if we set up a call, recorded that and turned that into a webinar?

**Reply by marcuslindblom** [reply]
@ayende that would be awesome but I'm on parental leave right now so I find it difficult to find time. It would be an amazing thing if you could set it up anyway. If you don't want to reveal the contents of a potential webinar I would appreciate an email with your solution on my problem.

Thanks for listening 😀

**Reply by ayende** [reply]
pinged your email

**Reply by georgiosd** [reply]
@marcuslindblom the way we do it is:
a) either store the full slug in the Post `post.Slug = "page-1/page-2"` when there is a 1-1 parent-child relationship that doesn't change or 

b) create `SlugReference` documents with the full slugs, pointing to the Post. You can have as many of these as you want, and the slug can (and should) be in the key: `slug-references/page-1/page-2` with `{ PostId: "posts/1" }` inside. Make sure you create these in a cluster-wide session so you can ensure uniqueness.

Of course I'd always welcome comments from Oren :)

**Reply by ayende** [reply]
about the reference, if you need that globally consistent, I would say use the compare exchange directly.

We then have a feature that allows you to do: `from Posts where id() = cmpxng('slug-references/page-1/page-2')`
